dev/DynamicDependency/PackageGraphManager.cpp

```diff
@@ -0,0 +1,241 @@
+    if (count)
+    {
+        *count = totalPackagesCount;
+    }
+
+    // Do we have anything to report?
+    if (totalPackagesCount == 0)
+    {
+        // No matches!
+        *bufferLength = 0;
+        return S_OK;
+    }
+
+    // Compute the buffer length used/needed and fill buffer (if we can)
+    auto bufferNeeded{ SerializePackageInfoToBuffer(flags, packagePathType, *bufferLength, buffer, matchingPackageInfo, dynamicPackagesCount, staticPackageInfo, staticPackagesCount) };
+    const auto isInsufficientBuffer{ *bufferLength < bufferNeeded };
+    *bufferLength = bufferNeeded;
+    RETURN_HR_IF(HRESULT_FROM_WIN32(ERROR_INSUFFICIENT_BUFFER), isInsufficientBuffer);
+
+    return S_OK;
+}
+CATCH_RETURN();
+
+UINT32 MddCore::PackageGraphManager::SerializePackageInfoToBuffer(
+    const UINT32 flags,
+    const PackagePathType packagePathType,
+    const UINT32 bufferLength,
+    BYTE* buffer,
+    const std::vector<const MddCore::PackageGraphNode*>& matchingPackageInfo,
+    const UINT32 dynamicPackagesCount,
+    const PACKAGE_INFO* staticPackageInfo,
+    const UINT32 staticPackagesCount)
+{
+    // Setup our initial pointers to write the PACKAGE_INFO[] and the variable length data
+    const auto totalPackagesCount{ dynamicPackagesCount + staticPackagesCount };
+    UINT32 bufferNeeded{ sizeof(PACKAGE_INFO) * totalPackagesCount };
+    auto toPackageInfo{ reinterpret_cast<PACKAGE_INFO*>(buffer) };
+
+    // Let's do it...
+    for (auto& matchingPackageGraphNode : matchingPackageInfo)
+    {
+        UINT32 fromPackagesCount{};
+        const void* fromPackagesBuffer{};
+        wil::unique_cotaskmem_ptr<BYTE[]> dynamicPackagesBuffer;
+
+        if (matchingPackageGraphNode)
+        {
+            // It's dynamic package data
+            fromPackagesCount = std::move(matchingPackageGraphNode->GetMatchingPackages(flags, packagePathType, dynamicPackagesBuffer));
+            fromPackagesBuffer = dynamicPackagesBuffer.get();
+        }
+        else
+        {
+            // It's static package data
+            fromPackagesCount = staticPackagesCount;
+            fromPackagesBuffer = staticPackageInfo;
+        }
+        auto fromPackageInfo{ static_cast<const PACKAGE_INFO*>(fromPackagesBuffer) };
+
+        // Copy the package info to the buffer
+        for (UINT32 index=0; index < fromPackagesCount; ++index, ++fromPackageInfo, ++toPackageInfo)
+        {
+            // Copy over the fixed length data
+            if (bufferNeeded <= bufferLength)
+            {
+                memset(toPackageInfo, 0, sizeof(*toPackageInfo));
+                toPackageInfo->flags = fromPackageInfo->flags;
+                toPackageInfo->packageId.processorArchitecture = fromPackageInfo->packageId.processorArchitecture;
+                toPackageInfo->packageId.version = fromPackageInfo->packageId.version;
+            }
+            SerializeStringToBuffer(bufferLength, buffer, bufferNeeded, toPackageInfo->path, fromPackageInfo->path);
```

Comment:
if bufferNeeded > bufferLength, it's OK to execute these SerializeStringToBuffer lines?

Reply:
Yes as `SerializeStringToBuffer` always computes the buffer needed but only copies data if the buffer's big enough. See line 234